opennebula/resource_opennebula_virtual_machine.go

```diff
@@ -1261,15 +1261,79 @@ func resourceOpennebulaVirtualMachineUpdate(d *schema.ResourceData, meta interfa
 	}
 
 	updateConf := false
+
+	// retrieve only template sections managed by updateconf method
 	tpl := vm.NewTemplate()
+	for _, name := range []string{"OS", "FEATURES", "INPUT", "GRAPHICS", "RAW", "CONTEXT"} {
+		vectors := vmInfos.Template.GetVectors(name)
+		for _, vec := range vectors {
+			tpl.Elements = append(tpl.Elements, vec)
+		}
+	}
 
 	if d.HasChange("os") {
 		updateConf = true
 
 		log.Printf("[DEBUG] Update os")
 
-		//Generate OS definition
-		addOS(tpl, d.Get("os").([]interface{}))
+		old, new := d.GetChange("os")
```

Comment:
it seems you use the same pattern for "os", "graphics" and "context", maybe use a function instead ?

Reply:
done